SPEC.md

```diff
@@ -0,0 +1,334 @@
+
+  **Body**: A JSON serialized [`OperationStarted`](#operationstarted) object.
+
+### Cancel Operation
+
+Request to cancel an operation.
+The operation may later complete as canceled or any other outcome.
+Handlers should ignore multiple cancelations of the same operation and return successfully if cancelation was already
+requested.
+
+#### Method
+
+`POST`
+
+#### Paths
+
+- `/api/v1/services/{service}/operations/{operation}/{operation_id}/cancel`
+
+#### Response Codes
+
+- `202 Accepted`: Cancelation request accepted.
+
+  **Body**: Unspecified.
+
+- `404 Not Found`: Operation ID not recognized or references deleted.
+
+  **Body**: Unspecified.
+
+### Get Operation Result
+
+Retrieve operation result.
+
+#### Method
+
+`GET`
+
+#### Paths
+
+- `/api/v1/services/{service}/operations/{operation}/{operation_id}/result`
+
+#### Query Parameters
+
+- `wait_deadline`: Optional. ISO 8601 date indicating the waiting period for a result.
+  If not specified, and the operation is still running, the request should resolve immediately with the current
+  operation status.
+
+#### Response Codes
+
+- `200 OK`: Operation completed successfully.
+
+  **Headers**:
+
+  - `Nexus-Operation-State: succeeded`
+
+  **Body**: Arbitrary data conveying the operation's result. Headers should specify content type and encoding.
+
+- `204 No Content`: Operation completed successfully with an empty result or is still running.
+
+  **Headers**:
+
+  - `Nexus-Operation-State: succeeded | running`
+
+  **Body**: Empty.
+
+- `482 Operation Failed`: Operation completed as `failed` or `canceled`.
+
+  **Headers**:
+
+  - `Content-Type: application/json`
+  - `Nexus-Operation-State: failed | canceled`
+
+  **Body**: A JSON serialized [`Failure`](#failure) object.
+
+- `408 Request Timeout`: The specified `wait_deadline` expired prior to operation completion.
+
+  The caller may re-issue this request to start a new long poll.
+
+  **Body**: Unspecified.
+
+- `404 Not Found`: Operation ID not recognized or references deleted.
+
+  **Body**: Unspecified.
+
+### Get Operation Info
+
+Retrieve operation details.
+
+#### Method
+
+`GET`
+
+#### Paths
+
+- `/api/v1/services/{service}/operations/{operation}/{operation_id}`
+
+#### Response Codes
+
+- `200 OK`: Successfully retrieved info.
+
+  **Headers**:
+
+  - `Content-Type: application/json`
+  - `Nexus-Operation-State: failed | canceled`
+
+  **Body**:
+
+  A JSON serialized [`OperationInfo`](#operationinfo) object.
+
+- `404 Not Found`: Operation ID not recognized or references deleted.
+
+## General Information on HTTP Response Codes
+
+The Nexus protocol follows standard HTTP practices:
+
+- `4xx` codes signify non-retryable framework-level errors.
+- `5xx` codes signify retryable framework-level errors.
+
+## Callback URLs
+
+Any HTTP URL can be used to deliver operation completions.
+
+Callers should ensure URLs contain sufficient information to correlate completions with initiators.
+
+For invoking a callback URL:
+
```

Comment:
You probably want to make clear that this is best effort, and is not retried forever until it returns a 200 OK
Also, it may be obvious that it uses a POST, but we could be explicit...

Reply:
I wonder whether restricting it to `POST` is unnecessarily limiting. Some clients might want to use `PUT` or `PATCH`, and I can also imagine some clients being in an unfortunate position where they are only able to specify a `GET` endpoint (due to something about their server framework, privileges, etc). Might we even consider other delivery methods? I've certainly implemented things in the past that had to be delivered by email on completion of a long-running operation.

Reply:
This is an intentionally limited API. It is the Nexus "protocol" we require servers to implement to talk to each other. A better more flexible HTTP API can be made on top (as can other protocol support). We don't want to support lots of loose implementations, that ambiguity can harm interoperability.

Reply:
I'll explicitly specify that we expect POST requests